File: strands-ts/src/background-tasks/in-process/__tests__/engine-test-helpers.ts
import { expect } from 'vitest'
import { InProcessTaskEngine } from '../engine.js'
import type { InterruptStateData } from '../../../interrupt.js'
import type { ToolResultBlockData } from '../../../types/messages.js'
import type { InProcessTaskEngineOptions, InProcessTaskRecord } from '../types.js'

type ExpectedTaskFields = Pick<InProcessTaskRecord, 'status'> &
  Partial<Pick<InProcessTaskRecord, 'state' | 'result' | 'failure'>>

export function createEngine(
  execute: InProcessTaskEngineOptions['execute'],
  options: Partial<InProcessTaskEngineOptions> = {}
): InProcessTaskEngine {
  return new InProcessTaskEngine({
    maxConcurrency: 2,
    timeout: 1_000,
    execute,
    onTaskUpdated: () => undefined,
    ...options,
  })
}

export function createAdmission(
  value: string
): Pick<InProcessTaskRecord, 'toolUseId' | 'toolName' | 'invocationStateId'> {
  return {
    toolUseId: value,
    toolName: value,
    invocationStateId: value,
  }
}

export function createResult(value: string): ToolResultBlockData {
  return {
    toolUseId: value,
    status: 'success',
    content: [{ text: value }],
  }
}

export function createState(value: string): InterruptStateData {
  return {
    interrupts: {
      [value]: {
        id: value,
        name: value,
      },
    },
    activated: true,
  }
}

export function getStateValue(state: InterruptStateData): string {
  return Object.keys(state.interrupts)[0]!
}

export function deferred<Value>(): { promise: Promise<Value>; resolve(value: Value): void } {
  let resolve!: (value: Value) => void
  const promise = new Promise<Value>((done) => {
    resolve = done
  })
  return { promise, resolve }
}

export function expectTask(
  actual: InProcessTaskRecord | undefined,
  task: InProcessTaskRecord,
  fields: ExpectedTaskFields
): void {
  expect(actual).toEqual({
    taskId: task.taskId,
    ...(task.idempotencyKey !== undefined && { idempotencyKey: task.idempotencyKey }),
    toolUseId: task.toolUseId,
    toolName: task.toolName,
    invocationStateId: task.invocationStateId,
    ...fields,
    createdAt: task.createdAt,
    updatedAt: expect.any(String),
  })
}

File: strands-py/tests/strands/background_tasks/in_process/_engine_test_helpers.py
from __future__ import annotations

import asyncio
from collections.abc import Awaitable, Callable
from typing import Any
from unittest.mock import ANY

from strands.background_tasks.in_process._engine import InProcessTaskEngine
from strands.background_tasks.in_process._types import (
    InProcessTaskExecutionContext,
    InProcessTaskExecutionOutcome,
    InProcessTaskRecord,
)
from strands.types.tools import ToolResult


def create_engine(
    execute: Callable[[InProcessTaskExecutionContext], Awaitable[InProcessTaskExecutionOutcome]],
    *,
    max_concurrency: int = 2,
    timeout: float = 1.0,
    on_task_updated: Callable[[InProcessTaskRecord], None] | None = None,
) -> InProcessTaskEngine:
    return InProcessTaskEngine(
        max_concurrency=max_concurrency,
        timeout=timeout,
        execute=execute,
        on_task_updated=on_task_updated or (lambda _: None),
    )


def create_admission(value: str) -> dict[str, str]:
    return {
        "tool_use_id": value,
        "tool_name": value,
        "invocation_state_id": value,
    }


def create_result(value: str) -> ToolResult:
    return {
        "toolUseId": value,
        "status": "success",
        "content": [{"text": value}],
    }


def create_state(value: str) -> dict[str, Any]:
    return {
        "interrupts": {
            value: {
                "id": value,
                "name": value,
            }
        },
        "activated": True,
    }


def get_state_value(state: dict[str, Any]) -> str:
    return next(iter(state["interrupts"]))


def assert_task(
    actual: InProcessTaskRecord | None,
    task: InProcessTaskRecord,
    fields: dict[str, Any],
) -> None:
    exp_task = {
        "task_id": task["task_id"],
        "tool_use_id": task["tool_use_id"],
        "tool_name": task["tool_name"],
        "invocation_state_id": task["invocation_state_id"],
        **fields,
        "created_at": task["created_at"],
        "last_updated_at": ANY,
    }
    assert actual == exp_task


def deferred() -> tuple[
    asyncio.Future[InProcessTaskExecutionOutcome],
    Callable[[InProcessTaskExecutionOutcome], None],
]:
    future = asyncio.get_running_loop().create_future()
    return future, future.set_result
